### PR TITLE
feat: Support keep-alive messages

### PR DESCRIPTION
**Pull request by mtojek**
Related: coder/coder#7581

This PR enables SSH server support for keep-alive messages. It introduces 2 more configuration options: `ClientAliveInterval` and `ClientAliveCountMax`.

This library is relatively old and not really covered with tests. I can add some if you find them beneficial.

How to test it (as CI doesn't start any tests):

Tab 1:

```
go run _examples/ssh-keepalive/keepalive.go
```

Tab 2:

```
ssh -o HostKeyAlgorithms=+ssh-rsa -vvv localhost -p 2222
```

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
...
+func KeepAliveRequestHandler(ctx Context, srv *Server, req *gossh.Request) (ok bool, payload []byte) {
+	log.Printf("Handle keep-alive request: %s (wantReply: %t)", req.Type, req.WantReply)
+
+	if !req.WantReply {
+		return true, nil
+	}
+
+	err := req.Reply(true, nil)
+	if err != nil {
+		log.Printf("Replying to client keep-alive request failed: %v", err)
```
Comment (johnstcn) [inline session.go]: What verbosity will this show as? Do we need to worry about coderd logs filling with keepalive messages?

File: session.go [hunk 2]
```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
...
-		case "subsystem":
-			if sess.handled {
-				req.Reply(false, nil)
-				continue
-			}
+		ctx.SetValue(ContextKeyKeepAliveCallback, keepAliveCallback)
+	} else {
+		// Configure a stopped ticker to prevent `<-keepAliveTicker.C` from panicking.
+		keepAliveTicker = time.NewTicker(math.MaxInt64)
+		keepAliveTicker.Stop()
```
Comment (johnstcn) [inline session.go]: Note for readers: `Stop()` does not close the channel

File: session.go [hunk 3]
```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
...
+	var keepAliveTicker *time.Ticker
+	if keepAliveEnabled {
+		keepAliveTicker = time.NewTicker(sess.keepAliveInterval)
+		defer keepAliveTicker.Stop()
 
-			sess.handled = true
-			req.Reply(true, nil)
+		keepAliveCallback = func() {
+			lastReceived = time.Now()
+			keepAliveTicker.Reset(sess.keepAliveInterval)
```
Comment (johnstcn) [inline session.go]: Does these need to be protected by a mutex?

File: _examples/ssh-keepalive/keepalive.go
Comment (johnstcn) [inline _examples/ssh-keepalive/keepalive.go:1]: Is it possible to add a test for this case? Start up a server with a low timeout, validate that 
1) Without keepalives, connection times out
2) With keepalives, connection stays up

File: session.go [hunk 4]
```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
...
+	var keepAliveCallback func()
 
-			// If there's a session policy callback, we need to confirm before
-			// accepting the session.
-			if sess.sessReqCb != nil && !sess.sessReqCb(sess, req.Type) {
-				sess.rawCmd = ""
-				req.Reply(false, nil)
-				continue
-			}
+	var keepAliveTicker *time.Ticker
```
Comment (mafredri) [inline session.go]: Suggestion: Perhaps a cleaner approach (to avoid creating a new stopped ticker) is to have `var keepAlive <-chan time.Time` instead and then `keepAlive = keepAliveTicker.C`. Since the nil channel blocks forever, there's no need for another assignment.

File: session.go [hunk 5]
```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
...
+				if err != nil {
+					log.Printf("Closing session failed: %v", err)
+				}
+				return
 			}
 
-			handler := sess.subsystemHandlers[payload.Value]
-			if handler == nil {
-				handler = sess.subsystemHandlers["default"]
+			log.Println("Send keep-alive request to the client")
```
Comment (mafredri) [inline session.go]: Have you confirmed logging is safe in these code paths and won't garble SSH sessions? (For instance, `ssh -vvv ...`.)

File: session.go (hunk 1 above)
Comment (mtojek) [reply session.go]: > What verbosity will this show as?

Once we update it in Coder, I guess that it will go with the default log level. 

> Do we need to worry about coderd logs filling with keepalive messages?

Actually, it depends on the Coder agent configuration. In theory, we should set the keep-alive ping close to the agent timeout. If the agent timeout is 120m, then we can set the interval to 119m.

File: session.go (hunk 2 above)
Comment (mtojek) [reply session.go]: Yes, that's the trick

File: session.go [hunk 6]
```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
...
+			if lastReceived.Add(time.Duration(sess.keepAliveCountMax) * sess.keepAliveInterval).Before(time.Now()) {
+				log.Println("Keep-alive reply not received. Close down the session.")
 
-			// If there's a session policy callback, we need to confirm before
-			// accepting the session.
-			if sess.sessReqCb != nil && !sess.sessReqCb(sess, req.Type) {
-				sess.rawCmd = ""
-				req.Reply(false, nil)
-				continue
+				err := sess.Close()
```
Comment (mafredri) [inline session.go]: Should we use `sess.Exit` instead? I believe `Close` directly calls an underlying channel. Not sure if there's a specific exit code that OpenSSH uses though.

File: session.go (hunk 3 above)
Comment (mtojek) [reply session.go]: I'm not quite sure as in the worst case two `Reset()` happens close to each other. We don't need perfect timing here.

Comment (mafredri) [reply session.go]: I would say yes, this is exposed on the context.

File: session.go [hunk 7]
```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
...
+				}
+				return
 			}
 
-			handler := sess.subsystemHandlers[payload.Value]
-			if handler == nil {
-				handler = sess.subsystemHandlers["default"]
+			log.Println("Send keep-alive request to the client")
+			// reply can be either false or true, but it always means that the client is alive
+			_, err := sess.SendRequest(keepAliveRequestType, true, nil)
```
Comment (mafredri) [inline session.go]: Since this is run synchronously, won't it hold up other SSH events? For instance, say a resize happens at the same time, that resize would be delayed by the whole RTT.

Perhaps we need a different kind of keepalive implementation? We probably can't simply do `go ...` here either since we want to ensure there's only one outbound at a time.

File: session.go [hunk 8]
```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
...
+			if err != nil {
+				log.Printf("Sending keep-alive request failed: %v", err)
+			} else {
+				log.Println("Client replied to keep-alive request.")
+				keepAliveCallback()
```
Comment (mafredri) [inline session.go]: Perhaps this should call `ctx.KeepAliveCallback()()`?

File: session.go [hunk 9]
```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
...
-			sess.Unlock()
-		default:
-			// TODO: debug log
-			req.Reply(false, nil)
 		}
+
+	}
+}
+
+func KeepAliveRequestHandler(ctx Context, srv *Server, req *gossh.Request) (ok bool, payload []byte) {
```
Comment (mafredri) [inline session.go:473]: It's a bit unclear to me who's sending requests and replies. My understanding is that server sends a keep-alive request, client responds, and we're done?

The prints in this function suggest that the client sends keep-alive requests too? Is this function called on both client and server or just one of them?

Comment (mtojek) [reply session.go:473]: > My understanding is that server sends a keep-alive request, client responds, and we're done?

This is only 1 part of the story.

Another mode is `ServerAliveInterval=13` when the client sends a keep-alive request every 13 seconds, and the SSH server needs to reply to it.

File: session.go (hunk 7 above)
Comment (mtojek) [reply session.go]: In reality, I expect fast replies if the client is up, and considering that we don't plan to send keep-alive requests too often (maybe every 2h?), I wouldn't complain considering our other networking problems.

I will take a look and see if this can be refactored into a non-blocking form. Perhaps, we need an extra channel for keep-alive replies/acks.

File: session.go (hunk 5 above)
Comment (mtojek) [reply session.go]: `ssh -vvv ....` is the client command, and we're printing logs on the server side, right? Please clarify as maybe I'm missing something.

File: session.go (hunk 7 above)
Comment (mafredri) [reply session.go]: Personally I don't think any added lag to the terminal is acceptable and will result in a degraded user experience. Proxied pings to my workspace take 300ms on dev.coder.com, if a keepalive+resize happen simultaneously I would expect that's around 450ms or more before the change is reflected in the terminal (an addition of 150ms is quite severe IMO).

File: session.go (hunk 6 above)
Comment (mtojek) [reply session.go]: I trust your experience here 👍  I don't know if it is better to call `sess.Exit(1)` than `sess.Close`.

File: session.go (hunk 8 above)
Comment (mtojek) [reply session.go]: Done

File: session.go (hunk 7 above)
Comment (mafredri) [reply session.go]: Actually, my math is off. It might be double: Resize from client 150ms -> KA from server 150ms -> KA Resp from client 150ms -> Resize on server + response 150ms -> Total 600ms.

File: session.go (hunk 5 above)
Comment (mafredri) [reply session.go]: You're right that this specific code is probably not problematic (except for maybe verbose logs on the agent?), but what about e.g. the handler? I'm assuming the client needs to handle keep-alives too? OpenSSH obviously does, but what about `coder ssh` in non-stdio mode?

Comment (mtojek) [reply session.go]: > I'm assuming the client needs to handle keep-alives too? OpenSSH obviously does, but what about coder ssh in non-stdio mode?

I don't know if default Go SSH client supports keep-alives, but if not, then most likely we will have to adjust the non-stdio code. I intend to focus on this once I'm done with changes in the library.

File: session.go (hunk 7 above)
Comment (mtojek) [reply session.go]: According to the source `pty-req` is also blocking, right?

```
	ok, err := s.ch.SendRequest("pty-req", true, Marshal(&req))
	if err == nil && !ok {
		err = errors.New("ssh: pty-req failed")
	}
```

so it seems that you can't intercept `pty-req` with `keepalive@openssh.com`, unless I understand it wrong. But I 
agree that aligning 2 consecutive blocking requests one after another can be concerning.

Comment (mafredri) [reply session.go]: Isn’t `pry-req` only done once at the start if the session? I think `window-change` is a more apt comparison.

Comment (mtojek) [reply session.go]: I missed that, thanks for clarification. Indeed, `window-change` is non-blocking, so it may be affected by keep-alive packets: 

```
_, err := s.ch.SendRequest("window-change", false, Marshal(&req))
return err
```

File: session.go [hunk 10]
```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
...
+	log.Printf("Handle keep-alive request: %s (wantReply: %t)", req.Type, req.WantReply)
+
+	if !req.WantReply {
+		return true, nil
+	}
+
+	err := req.Reply(true, nil)
```
Comment (code-asher) [inline session.go]: It looks like `req.Reply` is already called with the value the handler returns so we end up getting two replies on every keepalive the client sends.  Not sure if this actually matters though.

**Review by code-asher — COMMENTED** [review]
I tried out both `ClientAliveInterval` and `ServerAliveInterval` and they are working wonderfully.

One weird thing I noticed is that openssh's server actually seems to reply with `SSH_MSG_REQUEST_FAILURE` (82) while we are now replying with `SSH_MSG_REQUEST_SUCCESS` (81).

Similarly, the openssh client replies with `SSH_MSG_CHANNEL_FAILURE` (100).

Seems like odd behavior on `openssh`'s part and the implementation here is better but I thought I would mention it in case there is some reason it is supposed to reply with failures (or maybe I am just doing something wrong).

File: session.go (hunk 10 above)
Comment (code-asher) [reply session.go]: Ah I guess this also means the `WantReply` behavior does not quite work as intended either.

Comment (mtojek) [reply session.go]: Good catch, Asher. It looks like we don't need to call `Reply` at all!

**Comment by mtojek** [comment]
> Seems like odd behavior on openssh's part and the implementation here is better but I thought I would mention it in case there is some reason it is supposed to reply with failures (or maybe I am just doing something wrong).

I had a similar observation. I tried with some other existing SSH servers, and my Mac SSH client behaves the same way - it replies with a failure.

TL;DR I added a comment paragraph describing the anomaly.

File: session.go (hunk 3 above)
Comment (mtojek) [reply session.go]: Done

File: session.go (hunk 4 above)
Comment (mtojek) [reply session.go]: Good call, added.

File: session.go (hunk 7 above)
Comment (mtojek) [reply session.go]: Ok, I modified the source to use a goroutine and `TryLock` to see if the `req-reply` flow is in progress. Let me know your thoughts!

File: _examples/ssh-keepalive/keepalive.go
Comment (mtojek) [reply _examples/ssh-keepalive/keepalive.go:1]: I managed to cover 2. with session stats, so we can ensure what was called and how many times.

Unfortunately, I can't address 1. easily as intercepting SSH connection on request level is not possible. Long story short, I don't know how to deny the SSH client from responding to keep-alive requests.

File: session.go [hunk 11]
```diff
@@ -259,158 +276,249 @@ func (sess *session) Break(c chan<- bool) {
...
+					sess.winch <- win
+				}
 				req.Reply(ok, nil)
-				continue
-			}
-			req.Reply(false, nil)
-		case "window-change":
-			if sess.pty == nil {
+			case agentRequestType:
+				// TODO: option/callback to allow agent forwarding
```
Comment (johnstcn) [inline session.go:445]: nit: create follow-up issue and add link

File: session.go [hunk 12]
```diff
@@ -259,158 +276,249 @@ func (sess *session) Break(c chan<- bool) {
...
+				ok := false
+				sess.Lock()
+				if sess.breakCh != nil {
+					sess.breakCh <- true
+					ok = true
+				}
+				req.Reply(ok, nil)
+				sess.Unlock()
+			default:
+				// TODO: debug log
```
Comment (johnstcn) [inline session.go:462]: nit: create follow-up issue and add link

File: session_test.go [hunk 13]
```diff
@@ -476,3 +484,101 @@ func TestBreakWithoutChanRegistered(t *testing.T) {
...
+		}
+
+		// Verify that...
+		require.Zero(t, srv.keepAliveRequestHandlerCalled.Load())  // client didn't send any keep-alive requests,
+		require.Equal(t, 100, sshSession.serverRequestedKeepAlive) //  server requested keep-alive replies
+	})
+
+	t.Run("Server terminates connection due to no keep-alive replies", func(t *testing.T) {
+		t.Parallel()
+		t.Skip("Go SSH client doesn't support disabling replies to keep-alive requests. We can't test it easily without mocking logic.")
```
Comment (johnstcn) [inline session_test.go:588]: 😢

File: session.go [hunk 14]
```diff
@@ -259,158 +276,249 @@ func (sess *session) Break(c chan<- bool) {
...
-			if sess.sessReqCb != nil && !sess.sessReqCb(sess, req.Type) {
-				sess.rawCmd = ""
-				req.Reply(false, nil)
-				continue
-			}
+	var keepAliveRequestInProgress sync.Mutex
+	for {
+		select {
+		case <-keepAliveCh:
+			if lastReceived.Add(time.Duration(sess.keepAliveCountMax) * sess.keepAliveInterval).Before(time.Now()) {
```
Comment (johnstcn) [inline session.go]:
```
==================
WARNING: DATA RACE
Read at 0x00c0005a4570 by goroutine 474:
  github.com/gliderlabs/ssh.(*session).handleRequests()
      /Users/cian/src/coder/ssh/session.go:310 +0x1868
  github.com/gliderlabs/ssh.DefaultSessionHandler()
      /Users/cian/src/coder/ssh/session.go:128 +0x36c
  github.com/gliderlabs/ssh.(*Server).HandleConn.func4()
      /Users/cian/src/coder/ssh/server.go:319 +0x84

Previous write at 0x00c0005a4570 by goroutine 490:
  github.com/gliderlabs/ssh.(*session).handleRequests.func1()
      /Users/cian/src/coder/ssh/session.go:294 +0x5c
  github.com/gliderlabs/ssh.(*session).handleRequests.func2()
      /Users/cian/src/coder/ssh/session.go:337 +0x218

Goroutine 474 (running) created at:
  github.com/gliderlabs/ssh.(*Server).HandleConn()
      /Users/cian/src/coder/ssh/server.go:319 +0x75c
  github.com/gliderlabs/ssh.(*Server).serveOnce()
      /Users/cian/src/coder/ssh/session_test.go:28 +0x15c
  github.com/gliderlabs/ssh.newTestSession.func1()
      /Users/cian/src/coder/ssh/session_test.go:74 +0x4c

Goroutine 490 (finished) created at:
  github.com/gliderlabs/ssh.(*session).handleRequests()
      /Users/cian/src/coder/ssh/session.go:325 +0x19d8
  github.com/gliderlabs/ssh.DefaultSessionHandler()
      /Users/cian/src/coder/ssh/session.go:128 +0x36c
  github.com/gliderlabs/ssh.(*Server).HandleConn.func4()
      /Users/cian/src/coder/ssh/server.go:319 +0x84
==================
```

File: session_test.go [hunk 15]
```diff
@@ -476,3 +484,101 @@ func TestBreakWithoutChanRegistered(t *testing.T) {
...
+		defer cleanup()
+
+		errChan := make(chan error, 5)
+		go func() {
+			errChan <- session.Run("")
+		}()
+
+		// Wait for client to reply to 100 keep-alive requests.
+		require.Eventually(t, func() bool {
+			return sshSession.keepAliveReplyReceived == 100
```
Comment (johnstcn) [inline session_test.go]:
```
==================
WARNING: DATA RACE
Read at 0x00c00044c8f8 by goroutine 481:
  github.com/gliderlabs/ssh.TestSessionKeepAlive.func2.4()
      /Users/cian/src/coder/ssh/session_test.go:566 +0x44
  github.com/stretchr/testify/assert.Eventually.func1()
      /Users/cian/go/pkg/mod/github.com/stretchr/testify@v1.8.4/assert/assertions.go:1852 +0x38

Previous write at 0x00c00044c8f8 by goroutine 474:
  github.com/gliderlabs/ssh.DefaultSessionHandler()
      /Users/cian/src/coder/ssh/session.go:115 +0x74
  github.com/gliderlabs/ssh.(*Server).HandleConn.func4()
      /Users/cian/src/coder/ssh/server.go:319 +0x84

Goroutine 481 (running) created at:
  github.com/stretchr/testify/assert.Eventually()
      /Users/cian/go/pkg/mod/github.com/stretchr/testify@v1.8.4/assert/assertions.go:1852 +0x288
  github.com/stretchr/testify/require.Eventually()
      /Users/cian/go/pkg/mod/github.com/stretchr/testify@v1.8.4/require/require.go:401 +0x90
  github.com/gliderlabs/ssh.TestSessionKeepAlive.func2()
      /Users/cian/src/coder/ssh/session_test.go:565 +0x3ac
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x180
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x40

Goroutine 474 (running) created at:
  github.com/gliderlabs/ssh.(*Server).HandleConn()
      /Users/cian/src/coder/ssh/server.go:319 +0x75c
  github.com/gliderlabs/ssh.(*Server).serveOnce()
      /Users/cian/src/coder/ssh/session_test.go:28 +0x15c
  github.com/gliderlabs/ssh.newTestSession.func1()
      /Users/cian/src/coder/ssh/session_test.go:74 +0x4c
==================
```

**Review by johnstcn — COMMENTED** [review]
I tried it out and it appears to work fine.
Found a couple of data races though.

File: session.go (hunk 11 above)
Comment (mtojek) [reply session.go:445]: Oh, I must have reformatted the code as it isn't my TODO. I can create an issue though.

File: session.go (hunk 12 above)
Comment (mtojek) [reply session.go:462]: Same here

Comment (johnstcn) [reply session.go:462]: Ah then please disregard!

File: session.go (hunk 11 above)
Comment (johnstcn) [reply session.go:445]: Ah then please disregard!

**Comment by mtojek** [comment]
This is a great finding, Cian. I forgot to run this with race flag and discovered a can of worms. It looks like using `ctx.KeepAliveCallback()` always causes a race condition with a session loop. It is just a bad design.

I'm afraid that I have to start it over 💀

```
Read at 0x00c00038ccc0 by goroutine 95:
  github.com/gliderlabs/ssh.(*sshContext).KeepAliveCallback()
      /Users/mtojek/code/ssh/context.go:162 +0x44
  github.com/gliderlabs/ssh.KeepAliveRequestHandler()
      /Users/mtojek/code/ssh/session.go:530 +0x4c
  github.com/gliderlabs/ssh.(*Server).handleRequests()
      /Users/mtojek/code/ssh/server.go:335 +0x104
  github.com/gliderlabs/ssh.(*Server).HandleConn.func3()
      /Users/mtojek/code/ssh/server.go:309 +0x58

Previous write at 0x00c00038ccc0 by goroutine 100:
  github.com/gliderlabs/ssh.(*sshContext).SetValue()
      /Users/mtojek/code/ssh/context.go:127 +0x70
  github.com/gliderlabs/ssh.(*session).handleRequests()
      /Users/mtojek/code/ssh/session.go:311 +0x330
  github.com/gliderlabs/ssh.DefaultSessionHandler()
      /Users/mtojek/code/ssh/session.go:129 +0x36c
  github.com/gliderlabs/ssh.(*Server).HandleConn.func4()
      /Users/mtojek/code/ssh/server.go:319 +0x84
```

File: session.go (hunk 14 above)
Comment (mtojek) [reply session.go]: Killed

File: session_test.go (hunk 15 above)
Comment (mtojek) [reply session_test.go]: Killed

**Comment by mtojek** [comment]
Ok, I refactored the code by extracting the `KeepAlive` part to a separate entity guarded with a mutex.

File: keepalive.go [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+	ska.m.Lock()
+	defer ska.m.Unlock()
+
+	// log.Println("ska.Reset()")
+
+	ska.metrics.keepAliveReplyReceived++
+
+	if ska.ticker != nil && !ska.closed {
+		ska.lastReceived = time.Now()
+		ska.ticker.Reset(ska.clientAliveInterval)
```
Comment (johnstcn) [inline keepalive.go:62]: ensure that `ska.clientAliveInterval > 0` otherwise this will panic

File: keepalive.go [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+
+	ska.metrics.serverRequestedKeepAlive++
+}
+
+func (ska *SessionKeepAlive) Reset() {
+	ska.m.Lock()
+	defer ska.m.Unlock()
+
+	// log.Println("ska.Reset()")
+
```
Comment (johnstcn) [inline keepalive.go]:
```suggestion
```

File: keepalive.go [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+	ska.Reset()
+}
+
+func (ska *SessionKeepAlive) ServerRequestedKeepAliveCallback() {
+	ska.m.Lock()
+	defer ska.m.Unlock()
+
+	// log.Println("ska.ServerRequestedKeepAliveCallback()")
+
```
Comment (johnstcn) [inline keepalive.go]:
```suggestion
```

File: keepalive.go [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+	"sync"
+	"time"
+)
+
+type SessionKeepAlive struct {
+	clientAliveInterval time.Duration
+	clientAliveCountMax int
+
+	ticker       *time.Ticker
+	tickerCh     <-chan time.Time
```
Comment (johnstcn) [inline keepalive.go:13]: Do we need a separate field for the chan?

File: keepalive.go (hunk 16 above)
Comment (mtojek) [reply keepalive.go:62]: That condition is in the `NewSessionKeepAlive` constructor. Do you think that we need one more here?

File: keepalive.go (hunk 19 above)
Comment (mtojek) [reply keepalive.go:13]: This is recommendation from https://github.com/coder/ssh/pull/3#discussion_r1221329714

File: keepalive.go (hunk 16 above)
Comment (johnstcn) [reply keepalive.go:62]: Missed that!

**Review by johnstcn — APPROVED** [review]

File: session.go (hunk 5 above)
Comment (mtojek) [reply session.go]: I limited prints.

File: session.go [hunk 20]
```diff
@@ -259,158 +265,218 @@ func (sess *session) Break(c chan<- bool) {
...
-			// If there's a session policy callback, we need to confirm before
-			// accepting the session.
-			if sess.sessReqCb != nil && !sess.sessReqCb(sess, req.Type) {
-				sess.rawCmd = ""
-				req.Reply(false, nil)
-				continue
+				err := sess.Exit(0)
+				if err != nil {
+					log.Printf("Session exit failed: %v", err)
+				}
```
Comment (mafredri) [inline session.go]: I was looking at what exit code we should use (thinking 255 might be more appropriate), when I saw something gnarly: https://github.com/coder/ssh/blob/04bb837133e11ad897cef16572b3b732f76c4cc4/session.go#L193-L196

Essentially, if there's an error writing the message, the session will never be closed. Perhaps it's best to return to calling `Close` directly, like you were doing originally?

```suggestion
				_ = sess.Close()
```

File: session.go [hunk 21]
```diff
@@ -259,158 +265,218 @@ func (sess *session) Break(c chan<- bool) {
...
 }
+
+// KeepAliveRequestHandler replies to periodic client keep-alive requests:
+// client: send packet: type 80 (SSH_MSG_GLOBAL_REQUEST)
+// client: receive packet: type 81 (SSH_MSG_REQUEST_SUCCESS)
+//
+// It differs from OpenSSH client replies to keep-alive requests:
+// client: receive packet: type 98 (SSH_MSG_CHANNEL_REQUEST)
+// client: client_input_channel_req: channel 0 rtype keepalive@openssh.com reply 1
+// client: send packet: type 100 (SSH_MSG_CHANNEL_FAILURE)
```
Comment (mafredri) [inline session.go]: Why do we use a different type compared to OpenSSH?

In my testing, client initiated keep alives are type `80 client -> <- 82 server` and server initiated ones are `98 server -> <- 100 client`. Would it not make sense to mimic this?

File: session.go [hunk 22]
```diff
@@ -259,158 +265,218 @@ func (sess *session) Break(c chan<- bool) {
...
+//
+// Apparently, OpenSSH client always replies with 100, but it does not matter
+// as the server considers it as alive (only the response status is ignored).
+func KeepAliveRequestHandler(ctx Context, srv *Server, req *gossh.Request) (ok bool, payload []byte) {
+	ctx.KeepAlive().RequestHandlerCallback()
```
Comment (mafredri) [inline session.go]: If this handler is registered without enabling keepalive, this code will panic? Should we protect against that?

File: session_test.go [hunk 23]
```diff
@@ -476,3 +482,109 @@ func TestBreakWithoutChanRegistered(t *testing.T) {
...
+			},
+			SessionRequestCallback: func(sess Session, requestType string) bool {
+				sshSession = sess.(*session)
+				return true
+			},
+		}
+		session, client, cleanup := newTestSession(t, srv, nil)
+		defer cleanup()
+
+		errChan := make(chan error, 5)
```
Comment (mafredri) [inline session_test.go]:
```suggestion
		errChan := make(chan error, 1)
```

File: keepalive.go [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+
+func (ska *SessionKeepAlive) Metrics() KeepAliveMetrics {
+	ska.m.Lock()
+	defer ska.m.Unlock()
+
+	kam := ska.metrics
+	return &kam
+}
+
+type keepAliveMetrics struct {
```
Comment (mafredri) [inline keepalive.go]: Should we just export this struct / fields directly vs having the indirection of an interface? Or do we plan to expand this to have self-updating state in the future?

The interface, to me, suggests that once you pull the interface via `.Metrics()`, you don't need to call it more times since it likely would be a `*keepAliveMetrics` where the methods always return up-to-date values. And that's not the case here.

File: session_test.go [hunk 25]
```diff
@@ -476,3 +482,109 @@ func TestBreakWithoutChanRegistered(t *testing.T) {
...
+				defer m.Unlock()
+
+				sshSession = sess.(*session)
+				return true
+			},
+		}
+		session, _, cleanup := newTestSession(t, srv, nil)
+		defer cleanup()
+
+		errChan := make(chan error, 5)
```
Comment (mafredri) [inline session_test.go]:
```suggestion
		errChan := make(chan error, 1)
```

**Review by mafredri — COMMENTED** [review]
All in all implementation logic seems solid, but some minor things came up.

I'm mainly concerned about us diverging from OpenSSH with the message types (see comment), why don't we mimic the behavior?

File: session.go (hunk 22 above)
Comment (mtojek) [reply session.go]: `KeepAlive` as structure is supposed to be always present, it is just not active (not ticking). Where did you see the missing reference?

File: session.go (hunk 21 above)
Comment (mtojek) [reply session.go]: I changed to `80 -> <- 82` to be aligned with OpenSSH, but I can't say whether 81 is wrong according to the protocol (didn't find this information).

File: session_test.go (hunk 23 above)
Comment (mtojek) [reply session_test.go]: Done

File: session_test.go (hunk 25 above)
Comment (mtojek) [reply session_test.go]: Done

File: session.go (hunk 20 above)
Comment (mtojek) [reply session.go]: 255 and `sess.Close` 👍  It could be an improvement, yes.

File: session.go (hunk 22 above)
Comment (mafredri) [reply session.go]: Ah, you're right. I missed that. I guess if we're looking at fringe cases, it would be possible to call this in `ConnCallback` and it would panic.

File: keepalive.go (hunk 24 above)
Comment (mtojek) [reply keepalive.go]: As long as `KeepAliveMetrics` uses primitive fields, we can just use the struct 👍

File: session.go (hunk 20 above)
Comment (mafredri) [reply session.go]: I was thinking we would remove `sess.Exit` entirely, because when I thought about it I'm not sure we should be telling an "exit-status" to the client. The command that was running didn't exit, we just terminated the connection 🤔.

File: session.go (hunk 22 above)
Comment (mtojek) [reply session.go]: Fixed, added a condition against nil

**Review by mafredri — APPROVED** [review]
